crates/circuit/src/dag_circuit.rs

```diff
@@ -3571,6 +3610,35 @@ impl DAGCircuit {
         }
     }
 
+    /// Returns an iterator of the ancestors indices of a node.
+    pub fn ancestors<'a>(&'a self, node: &DAGNode) -> impl Iterator<Item = NodeIndex> + 'a {
```

Comment:
For all of these Rust-native methods, we ought to deal in terms of `NodeIndex` for both the inputs and the outputs, i.e. you should change `node` to be a `NodeIndex`.

The reason being that `DAGNode` is really only a Python concept, since a Rust caller would only ever have a `DAGNode` if they were working with Python.